qcodes/instrument_drivers/Keysight/N9030B.py

```diff
@@ -0,0 +1,529 @@
+
+
+class N9030B(VisaInstrument):
+    """
+    Driver for Keysight N9030B PXA signal analyzer.
+    """
+
+    def __init__(self, name: str, address: str, **kwargs: Any) -> None:
+        super().__init__(name, address, terminator='\n', **kwargs)
+
+        self._min_freq: float
+        self._max_freq: float
+        self._additional_wait: float = 1
+
+        self.add_parameter(
+            name="mode",
+            get_cmd=":INSTrument:SELect?",
+            set_cmd=":INSTrument:SELect {}",
+            vals=Enum(*self._available_modes()),
+            docstring="Allows setting of different modes present and licensed "
+                      "for the instrument."
+        )
+
+        self.add_parameter(
+            name="measurement",
+            get_cmd=":CONFigure?",
+            set_cmd=":CONFigure:{}",
+            vals=Enum(*self._available_meas()),
+            docstring="Sets measurement type from among the available "
+                      "measurement types."
+        )
+
+        self.add_parameter(
+            name="cont_meas",
+            initial_value=False,
+            get_cmd=":INITiate:CONTinuous?",
+            set_cmd=self._enable_cont_meas,
+            val_mapping=create_on_off_val_mapping(on_val="ON", off_val="OFF"),
+            docstring="Enables or disables continuous measurement."
+        )
+
+        self.add_parameter(
+            name="format",
+            get_cmd=":FORMat:TRACe:DATA?",
+            set_cmd=":FORMat:TRACe:DATA {}",
+            val_mapping={
+                "ascii": "ASCii",
+                "int32": "INTeger,32",
+                "real32": "REAL,32",
+                "real64": "REAL,64"
+            },
+            docstring="Sets up format of data received"
+        )
+        self.connect_message()
+
+    def _available_modes(self) -> Tuple[str, ...]:
+        """
+        Returns present and licensed modes for the instrument.
+        """
+        available_modes = self.ask(":INSTrument:CATalog?")
+        return tuple(available_modes.split(','))
+
+    def _available_meas(self) -> Tuple[str, ...]:
+        """
+        Gives available measurement with a given mode for the instrument
+        """
+        available_meas = self.ask(":CONFigure:CATalog?")
+        return tuple(available_meas.split(','))
+
+    def _enable_cont_meas(self, val: str) -> None:
+        """
+        Sets continuous measurement to ON or OFF.
+        """
+        self.write(f":INITiate:CONTinuous {val}")
+
+    def _options(self) -> Tuple[str, ...]:
+        """
+        Returns installed options numbers.
+        """
+        options_raw = self.ask('*OPT?')
+        return tuple(options_raw.split(','))
+
+    def reset(self) -> None:
+        """
+        Reset the instrument by sending the RST command
+        """
+        self.write("*RST")
+
+    def abort(self) -> None:
+        """
+        Aborts the measurement
+        """
+        self.write(":ABORt")
+
+
+class SpectrumAnalyzer(N9030B):
```

Comment:
What was the motivation for breaking up the modes into separate classes like this?

Reply:
Just to bring over our discussion, I suggest not breaking up SpectrumAnalyzer and PhaseNoise this way. Consider using InstrumentChannel and adding these modes as submodules of the N9030B instrument. Let me know if you want more clarification or help!

Reply:
Sure, for now I am good. Will update these.